amazon-kinesis-client/src/test/java/software/amazon/kinesis/lifecycle/ShardConsumerSubscriberTest.java

```diff
@@ -311,6 +310,64 @@ public void restartAfterRequestTimerExpiresTest() throws Exception {
 
     }
 
+    @Test
+    public void restartAfterRequestTimerExpiresAfterInitialSubscriptionFailsTest() throws Exception {
+
+        executorService = Executors.newFixedThreadPool(1, new ThreadFactoryBuilder()
+                .setNameFormat("test-" + testName.getMethodName() + "-%04d").setDaemon(true).build());
+
+        // Mock record publisher which doesn't publish any records on first try which simulates any scenario which
+        // causes first subscription try to fail.
+        recordsPublisher = new RecordPublisherWithInitialFailureSubscription();
+        subscriber = new ShardConsumerSubscriber(recordsPublisher, executorService, bufferSize, shardConsumer, 0);
+        addUniqueItem(1);
+
+        List<ProcessRecordsInput> received = new ArrayList<>();
+        doAnswer(a -> {
+            ProcessRecordsInput input = a.getArgumentAt(0, ProcessRecordsInput.class);
+            received.add(input);
+            if (input.records().stream().anyMatch(r -> StringUtils.startsWith(r.partitionKey(), TERMINAL_MARKER))) {
+                synchronized (processedNotifier) {
+                    processedNotifier.notifyAll();
+                }
+            }
+            return null;
+        }).when(shardConsumer).handleInput(any(ProcessRecordsInput.class), any(Subscription.class));
+
+        // First try to start subscriptions.
+        synchronized (processedNotifier) {
+            subscriber.startSubscriptions();
+        }
+
+        // Verifying that there are no interactions with shardConsumer mock indicating no records were sent back and
+        // subscription has not started correctly.
+        verify(shardConsumer, never()).handleInput(argThat(eqProcessRecordsInput(processRecordsInput)),
+                any(Subscription.class));
+
+        Stream.iterate(2, i -> i + 1).limit(98).forEach(this::addUniqueItem);
+
+        addTerminalMarker(2);
+
+        // Doing the health check to allow the subscription to restart.
+        assertThat(subscriber.healthCheck(1), nullValue());
+
+        // Allow time for processing of the records to end in the executor thread which call notifyAll as it gets the
+        // terminal record.
+        synchronized (processedNotifier) {
+            processedNotifier.wait(500);
```

Comment:
Is 500 millis guaranteed to be enough time to reach the terminal record, even under conditions of high CPU usage or other potential slowdowns? We don't want a test failure based on internals we have no control over. 

Would mocking your executorService and telling it explicitly what to do on each call let you avoid this wait?

Reply:
The test here is actually about sending records from RecordPublishers to the Subscriber through the executor service. The avg time taken to receive all the records from the record publisher is around 60-80 ms hence 500 seems to be a good general timeout. I can update it to a higher value since this is a timeout and having it higher should not affect our test run time.